### PR TITLE
Remove BoostRandom and clean a bit the trace of boost

### PR DESCRIPTION
**Pull request by ye-luo**
Marked draft. Wait for testing results of StdRandom switch.

## Proposed changes
Remove BoostRandom and clean a bit the trace of boost

## What type(s) of changes does this code introduce?
- Other (please describe): Remove no more needed code

### Does this introduce a breaking change?
- Yes. No more access to boost RNG.

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes/No. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/Utilities/tests/test_rng.cpp [hunk 1]
```diff
@@ -22,28 +21,6 @@ using std::string;
 using std::vector;
 namespace qmcplusplus
 {
-#ifdef HAVE_LIBBOOST
```
Comment (PDoakORNL) [inline src/Utilities/tests/test_rng.cpp:25]: Std::mt19937 should still be deterministic with default construction.  Or at least that's what I think specifying what the 10000 output should guarantee.

**Review by PDoakORNL — DISMISSED** [review]
I think it would be good to make sure the mt19937 is deterministic with default construction. Although I am fine with flagging this as an issue to be taken up after clearing #3700

File: src/Utilities/tests/test_rng.cpp (hunk 1 above)
Comment (ye-luo) [reply src/Utilities/tests/test_rng.cpp:25]: I will add the determinism to the StdRandom test.

**Comment by prckent** [comment]
I had to restart the long tests since I didn't change the default timeout.

File: src/Utilities/tests/test_rng.cpp (hunk 1 above)
Comment (ye-luo) [reply src/Utilities/tests/test_rng.cpp:25]: Just migrated the test from BoostRandom to StdRandom. Kept testing only the first 3 elements. Even if we check 10000th and it went wrong, there won't be much we can do. So testing first 3 should be sufficient.

**Comment by prckent** [comment]
Test results with C++RNG (GCC+MKL real build):
```
97% tests passed, 70 tests failed out of 2098

Label Time Summary:
QMCPACK                     = 1690688.44 sec*proc (256 tests)
QMCPACK-checking-results    = 171.58 sec*proc (1491 tests)
abort                       = 84076.38 sec*proc (24 tests)
bug                         = 375510.87 sec*proc (107 tests)
cause_unknown               = 176537.99 sec*proc (143 tests)
check_fail                  = 110430.65 sec*proc (23 tests)
converter                   =  21.83 sec*proc (20 tests)
coverage                    = 399.16 sec*proc (5 tests)
definite_stat_fail          = 172256.39 sec*proc (46 tests)
deterministic               = 1087.87 sec*proc (1031 tests)
hard_fail                   = 8753.36 sec*proc (20 tests)
intermittent_stat_fail      = 245071.25 sec*proc (169 tests)
nexus                       =  69.55 sec*proc (70 tests)
performance                 = 12671.84 sec*proc (78 tests)
poor_test                   =  16.78 sec*proc (14 tests)
quality_unknown             = 1927301.24 sec*proc (1979 tests)
unit                        =  59.43 sec*proc (36 tests)
unstable                    = 115.79 sec*proc (716 tests)

Total Test time (real) = 60776.08 sec

The following tests FAILED:
        338 - short-LiH_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check (Failed)
        340 - long-LiH_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check (Failed)
        368 - long-LiH_ae_qp-vmc_hf_noj-1-16-totenergy (Failed)
        449 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4 (Failed)
        450 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4-check (Failed)
        451 - long-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4 (Failed)
        452 - long-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4-check (Failed)
        454 - short-LiH_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check (Failed)
        456 - long-LiH_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check (Failed)
        472 - short-H4-RHF-orb-opt-16-1-totenergy (Failed)
        476 - short-H4-orb-opt-1-16-totenergy (Failed)
        477 - short-H4-orb-opt-dmc-16-1 (Failed)
        478 - short-H4-orb-opt-dmc-16-1-totenergy (Failed)
        480 - short-H4-orb-opt-dmc-1-16-totenergy (Failed)
        501 - short-H4-opt-OneShiftOnly-1-16 (Timeout)
        504 - short-H4-msd-orb-opt-16-1-totenergy (Failed)
        557 - short-LiH_ae-vmc_msdj_noj-1-16-variance (Failed)
        739 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1 (Failed)
        740 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-totenergy_A (Failed)
        741 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-totenergy_B (Failed)
        742 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-dtotenergy_AB (Failed)
        743 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-ionion_A (Failed)
        744 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-ionion_B (Failed)
        745 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-dionion_AB (Failed)
        746 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-eeenergy_A (Failed)
        747 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-eeenergy_B (Failed)
        748 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-deeenergy_AB (Failed)
        749 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1 (Failed)
        750 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-totenergy_A (Failed)
        751 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-totenergy_B (Failed)
        752 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-dtotenergy_AB (Failed)
        753 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-ionion_A (Failed)
        754 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-ionion_B (Failed)
        755 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-dionion_AB (Failed)
        756 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-eeenergy_A (Failed)
        757 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-eeenergy_B (Failed)
        758 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-deeenergy_AB (Failed)
        759 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1 (Failed)
        760 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-totenergy_A (Failed)
        761 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-totenergy_B (Failed)
        762 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-dtotenergy_AB (Failed)
        763 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-ionion_A (Failed)
        764 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-ionion_B (Failed)
        765 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-dionion_AB (Failed)
        766 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-eeenergy_A (Failed)
        767 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-eeenergy_B (Failed)
        768 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-deeenergy_AB (Failed)
        818 - short-diamondC_1x1x1_pp-vmc_sdj-meshf-1-16-nonlocalecp (Failed)
        948 - short-diamondC_1x1x1_pp-vmcbatch-estimator-onebodydensitymatrices-4-4-check (Failed)
        956 - short-diamondC_1x1x1_pp-vmc-noJ-estimator-1rdm-4-4-check (Failed)
        958 - long-diamondC_1x1x1_pp-vmc-noJ-estimator-1rdm-4-4-check (Failed)
        960 - short-diamondC_1x1x1_pp-vmc-J2-estimator-1rdm-4-4-check (Failed)
        962 - long-diamondC_1x1x1_pp-vmc-J2-estimator-1rdm-4-4-check (Failed)
        968 - short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check (Failed)
        970 - long-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check (Failed)
        974 - long-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check (Failed)
        1776 - long-diamondC_2x1x1-Gaussian_pp_MSD-1-16-totenergy (Failed)
        1778 - long-diamondC_2x1x1-Gaussian_pp_MSD-1-16-potential (Failed)
        2086 - ntest_nexus_nxs_sim (Failed)
        2087 - ntest_nexus_qmca (Failed)
        2088 - ntest_nexus_qmc_fit (Failed)
        2089 - ntest_nexus_qdens (Failed)
        2091 - ntest_nexus_example_gamess_H2O (Failed)
        2092 - ntest_nexus_example_pwscf_relax_Ge_T (Failed)
        2093 - ntest_nexus_example_qmcpack_H2O (Failed)
        2094 - ntest_nexus_example_qmcpack_LiH (Failed)
        2095 - ntest_nexus_example_qmcpack_c20 (Failed)
        2096 - ntest_nexus_example_qmcpack_diamond (Failed)
        2097 - ntest_nexus_example_qmcpack_graphene (Failed)
        2098 - ntest_nexus_example_qmcpack_oxygen_dimer (Failed)
Errors while running CTest
```

Test results with boost RNG:
```
96% tests passed, 75 tests failed out of 2098

Label Time Summary:
QMCPACK                     = 1640163.17 sec*proc (256 tests)
QMCPACK-checking-results    = 132.97 sec*proc (1491 tests)
abort                       = 92556.92 sec*proc (24 tests)
bug                         = 349859.14 sec*proc (107 tests)
cause_unknown               = 170524.24 sec*proc (143 tests)
check_fail                  = 96801.10 sec*proc (23 tests)
converter                   =  19.81 sec*proc (20 tests)
coverage                    = 401.15 sec*proc (5 tests)
definite_stat_fail          = 151586.35 sec*proc (46 tests)
deterministic               = 1030.45 sec*proc (1031 tests)
hard_fail                   = 8921.67 sec*proc (20 tests)
intermittent_stat_fail      = 250130.86 sec*proc (169 tests)
nexus                       =  79.94 sec*proc (70 tests)
performance                 = 11934.06 sec*proc (78 tests)
poor_test                   =  18.95 sec*proc (14 tests)
quality_unknown             = 1827350.82 sec*proc (1979 tests)
unit                        =  85.86 sec*proc (36 tests)
unstable                    =  76.32 sec*proc (716 tests)

Total Test time (real) = 57691.89 sec

The following tests FAILED:
        321 - short-C4_md-ae-ground-qp-vmc-1-16-kinetic (Failed)
        325 - short-C4_md-ae-ground-qp-vmc-1-16-potential (Failed)
        338 - short-LiH_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check (Failed)
        340 - long-LiH_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check (Failed)
        367 - long-LiH_ae_qp-vmc_hf_noj-1-16-kinetic (Failed)
        368 - long-LiH_ae_qp-vmc_hf_noj-1-16-totenergy (Failed)
        449 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4 (Failed)
        450 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4-check (Failed)
        451 - long-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4 (Failed)
        452 - long-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4-check (Failed)
        454 - short-LiH_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check (Failed)
        456 - long-LiH_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check (Failed)
        472 - short-H4-RHF-orb-opt-16-1-totenergy (Failed)
        476 - short-H4-orb-opt-1-16-totenergy (Failed)
        477 - short-H4-orb-opt-dmc-16-1 (Failed)
        478 - short-H4-orb-opt-dmc-16-1-totenergy (Failed)
        480 - short-H4-orb-opt-dmc-1-16-totenergy (Failed)
        504 - short-H4-msd-orb-opt-16-1-totenergy (Failed)
        639 - short-LiH_solid_1x1x1_pp-gamma-vmc_hf_noj-1-16-localecp (Failed)
        739 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1 (Failed)
        740 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-totenergy_A (Failed)
        741 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-totenergy_B (Failed)
        742 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-dtotenergy_AB (Failed)
        743 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-ionion_A (Failed)
        744 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-ionion_B (Failed)
        745 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-dionion_AB (Failed)
        746 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-eeenergy_A (Failed)
        747 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-eeenergy_B (Failed)
        748 - short-bccH_1x1x1_ae-csvmc-pbyp-nodrift_sdj-16-1-deeenergy_AB (Failed)
        749 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1 (Failed)
        750 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-totenergy_A (Failed)
        751 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-totenergy_B (Failed)
        752 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-dtotenergy_AB (Failed)
        753 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-ionion_A (Failed)
        754 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-ionion_B (Failed)
        755 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-dionion_AB (Failed)
        756 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-eeenergy_A (Failed)
        757 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-eeenergy_B (Failed)
        758 - short-bccH_1x1x1_ae-csvmc-all-nodrift_sdj-16-1-deeenergy_AB (Failed)
        759 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1 (Failed)
        760 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-totenergy_A (Failed)
        761 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-totenergy_B (Failed)
        762 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-dtotenergy_AB (Failed)
        763 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-ionion_A (Failed)
        764 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-ionion_B (Failed)
        765 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-dionion_AB (Failed)
        766 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-eeenergy_A (Failed)
        767 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-eeenergy_B (Failed)
        768 - short-bccH_1x1x1_ae-csvmc-all_sdj-16-1-deeenergy_AB (Failed)
        772 - short-bccH_1x1x1_ae-rmc_sdj-4-4-totenergy (Failed)
        916 - long-diamondC_1x1x1_pp-vmc-dmc-allp_sdj-1-16-1-nonlocalecp (Failed)
        948 - short-diamondC_1x1x1_pp-vmcbatch-estimator-onebodydensitymatrices-4-4-check (Failed)
        956 - short-diamondC_1x1x1_pp-vmc-noJ-estimator-1rdm-4-4-check (Failed)
        958 - long-diamondC_1x1x1_pp-vmc-noJ-estimator-1rdm-4-4-check (Failed)
        960 - short-diamondC_1x1x1_pp-vmc-J2-estimator-1rdm-4-4-check (Failed)
        962 - long-diamondC_1x1x1_pp-vmc-J2-estimator-1rdm-4-4-check (Failed)
        964 - short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check (Failed)
        968 - short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check (Failed)
        974 - long-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check (Failed)
        1342 - long-diamondC_2x1x1_pp-vmc_sdj-1-16-flux (Failed)
        1776 - long-diamondC_2x1x1-Gaussian_pp_MSD-1-16-totenergy (Failed)
        1778 - long-diamondC_2x1x1-Gaussian_pp_MSD-1-16-potential (Failed)
        1940 - long-monoO_1x1x1_pp-vmc_sdj-1-16-flux (Failed)
        2086 - ntest_nexus_nxs_sim (Failed)
        2087 - ntest_nexus_qmca (Failed)
        2088 - ntest_nexus_qmc_fit (Failed)
        2089 - ntest_nexus_qdens (Failed)
        2091 - ntest_nexus_example_gamess_H2O (Failed)
        2092 - ntest_nexus_example_pwscf_relax_Ge_T (Failed)
        2093 - ntest_nexus_example_qmcpack_H2O (Failed)
        2094 - ntest_nexus_example_qmcpack_LiH (Failed)
        2095 - ntest_nexus_example_qmcpack_c20 (Failed)
        2096 - ntest_nexus_example_qmcpack_diamond (Failed)
        2097 - ntest_nexus_example_qmcpack_graphene (Failed)
        2098 - ntest_nexus_example_qmcpack_oxygen_dimer (Failed)
Errors while running CTest
```

nexus failures are from the ongoing python problems on sulfur.

As expected, nearly all the tests pass and those that fail are stochastic and not obviously different from those run with the boost RNG. Therefore we can delete the old boost RNG implementation.

**Review by prckent — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**